dowhy/causal_identifiers/IDIdentifier.py

```diff
@@ -0,0 +1,260 @@
+        node_idx_list.sort()
+        adjacency_matrix_induced = adjacency_matrix.copy()
+        adjacency_matrix_induced = adjacency_matrix_induced[node_idx_list]
+        adjacency_matrix_induced = adjacency_matrix_induced[:, node_idx_list]
+        return adjacency_matrix_induced        
+
+    def _find_c_components(self, adjacency_matrix, node_set, idx2node):
+
+        if node_set is None:
+            node_set = self._node_names
+        num_nodes = len(node_set)
+        adj_matrix = adjacency_matrix.copy()
+        adjacency_list = [[] for _ in range(num_nodes)]
+
+        # Modify graph such that it only contains bidirected edges
+        for h in range(0, num_nodes-1):
+            for w in range(h+1, num_nodes):
+                if adjacency_matrix[h, w]==1 and adjacency_matrix[w, h]==1:
+                    adjacency_list[h].append(w)
+                    adjacency_list[w].append(h)
+                else:
+                    adj_matrix[h, w] = 0
+                    adj_matrix[w, h] = 0
+
+        # Find c components by finding connected components on the undirected graph
+        visited = [False for _ in range(num_nodes)]
+
+        def dfs(node_idx, component):
+            visited[node_idx] = True
+            component.add(idx2node[node_idx])
+            for neighbour in adjacency_list[node_idx]:
+                if visited[neighbour] == False:
+                    dfs(neighbour, component)
+
+        c_components = []
+        for i in range(num_nodes):
+            if visited[i] == False:
+                component = OrderedSet()
+                dfs(i, component)
+                c_components.append(component)
+
+        return c_components
+    
+    def _idx_node_mapping(self, node_names):
+        node2idx = {}
+        idx2node = {}
+        for i, node in enumerate(node_names):
+            node2idx[node] = i
+            idx2node[i] = node
+        return node2idx, idx2node
+
+    # def _is_present(self, estimator):
+    #     string = ""
+    #     for node in estimator['outcome_vars']:
+    #         string += node
+    #     string += "|"
+    #     for node in estimator['condition_vars']:
+    #         string += node
+    #     string += "|"
+    #     for node in estimator['marginalize_vars']:
+    #         string += node
+    
+    #     if string not in self._estimator_set:
+    #         self._estimator_set.add(string)
+    #         return False
+    #     return True
+
+    def _print_estimator(self, estimator, level):
+        string = ""
+        if isinstance(estimator, IDExpression):
+            # string += "List:\n"
+            for i, expression in enumerate(estimator.get_val(return_type='prod')):
+                string += self._print_estimator(expression, level+1)
+        else:
+            for i in range(level):
+                string += "\t"
+            string += str(estimator)
+            string += "\n"
+        return string
+
+    def __str__(self):
```

Comment:
Ideally, this str function should call the IDExpression's str function recursively.